src/ntfs/write.rs

```diff
@@ -239,181 +215,250 @@ struct Fares<'a> {
 
-        if perimeters.is_empty() {
+        if included_lines.len() + included_networks.len() == 0 {
             warn!(
-                "Failed to find TicketUsePerimeter for TicketUse {:?}",
-                ticket_use_id
+                "The ticket_use_id {} is ignored since it has no included line or network, \
+                 and at least one must exists for a ticket_use_id to be valid.",
+                ticket_use.id
             );
             continue;
         }
 
-        let prices = get_prices(fares.ticket_prices, &ticket.id);
-        if prices.is_empty() {
-            warn!("Failed to find TicketPrice for Ticket {:?}", ticket.id);
-            continue;
-        }
-
-        for price in prices {
-            prices_v1.insert(PriceV1::try_from((ticket, price))?);
-        }
-
-        for perimeter in perimeters {
-            fares_v1.insert(FareV1 {
-                before_change: "*".to_string(),
-                after_change: format!("line=line:{}", perimeter.object_id),
-                start_trip: format!("stoparea=stop_area:{}", ticket_use_restriction.use_origin),
-                end_trip: format!(
-                    "stoparea=stop_area:{}",
-                    ticket_use_restriction.use_destination
-                ),
-                global_condition: String::new(),
-                ticket_id: ticket.id.clone(),
-            });
-        }
-    }
-    Ok(())
-}
-
-// Conversion of a flat fare on a specific network
-// https://github.com/CanalTP/transit_model/blob/master/src/documentation/ntfs_fare_conversion_v2_to_V1.md#conversion-of-a-flat-fare-on-a-specific-network
-fn insert_flat_fare_as_fare_v1(
-    fares: &Fares,
-    prices_v1: &mut BTreeSet<PriceV1>,
-    fares_v1: &mut BTreeSet<FareV1>,
-) -> Result<()> {
-    let ticket_use_perimeters = fares
-        .ticket_use_perimeters
-        .values()
-        .filter(|p| p.object_type == ObjectType::Network)
-        .filter(|p| p.perimeter_action == PerimeterAction::Included)
-        .filter_map(|ticket_use_perimeter| {
-            fares
-                .ticket_uses
-                .get(&ticket_use_perimeter.ticket_use_id)
-                .and_then(|ticket_use| fares.tickets.get(&ticket_use.ticket_id))
-                .map(|ticket| (ticket, ticket_use_perimeter))
-                .or_else(|| {
-                    warn!(
-                        "Failed to find Ticket for TicketUseId {:?}",
-                        ticket_use_perimeter.ticket_use_id
-                    );
-                    None
-                })
-        });
-    for (ticket, ticket_use_perimeter) in ticket_use_perimeters {
-        let prices = get_prices(fares.ticket_prices, &ticket.id);
-        if prices.is_empty() {
-            warn!("Failed to find TicketPrice for Ticket {:?}", ticket.id);
-            continue;
-        }
+        // Now the restrictions for our ticket_use_id
+        let restrictions: Vec<&TicketUseRestriction> = fares
```

Comment:
On each ticket, the iteration is realized on the entirety of `fares.ticket_use_restriction`. There might be a way to be more efficient by storing these relations in the `Model` and then use `get_corresponding()` instead of the current iteration. We can discuss it if you want.

Reply:
@woshilapin be careful not to create short circuits by adding relations.
You can break everything

Reply:
Yes, fares are not in graph

Reply:
I mean if we add more relations, get_corresponding may not work as before. the paths between nodes in the graph may change and result a bad behaviour

Reply:
There is still a lot of work to do on fares v2 (checking cardinalities and relations between the objects in the 5 collections themselves, and others, stopareas, lines, networks, etc.). We have a ticket in the backlog.

Reply:
In fact, we already had this debate there. hence the existence of the ticket in the backlog :grinning:

Reply:
@woshilapin if you go down this way then, for each `ticket_use` :

- the whole `fares.ticket_use_perimeters` is scanned in `extract_perimeters_for_ticket_use(..)`
- the whole `fares.ticket_use_restrictions` is scanned as you point
- the whole `fares.ticket_prices`  is also scanned

Correct me if I'm wrong, by it seems to me that using `get_corresponding` is doing a `filter_map`, which is more or less equivalent to what happens here (?)

I think it may be possible to scan each file only once, and incrementally build a data structure as below, and then generate the output. 
Not sure if it is worth the effort though !

```rust 

struct MyFaresV2 {
    ticket_uses : HashMap<String,TicketUse> // maps a ticket_use_id to its data
    ticket_prices : HashMap<String, Vec<Price> > // maps a ticket_id to its prices
}

struct Price {
    currency : String,
    price : Decimal,
    validity_start : Date,
    validity_end : Date, 
}

struct TicketUse {
    id : String,
    ticket_id : String,
    ticket_name : String,
    ticket_comment : String,
    max_transfers : i64,
    boarding_time_limit : i64,
    alighting_time_limit : i64,
    included_networks : Vec<String>,
    included_lines : Vec<String>,
    excluded_lines : Vec<String>,
    restrictions : Vec<Restriction>,
}

enum Restriction {
    Zone(String, String),
    OD(String, String)
}

```

Reply:
What you propose is pretty much what is doing `get_corresponding` when creating the object `Model`. When `Model` is built, all relations are created once.

That said, we discussed it in daily this morning and we might address this refactorization later. For now, let's not put more complexity to the current PR.